tests/run/cpython_capi.pyx

```diff
@@ -30,16 +48,25 @@ def test_pymalloc_raw():
     >>> test_pymalloc_raw()
     3
     """
+    cdef short i
+    cdef short* s
     cdef char* m
     cdef char* m2 = NULL
     with nogil:
+        s = <short*> mem.PyMem_RawCalloc(10, sizeof(*s))
```

Comment:
In order to avoid calling into code that isn't nogil-safe, what about
```
if cpython.version.PY_VERSION_HEX >= 0x03050000:
    with nogil:
        … mem.PyMem_RawCalloc(…)
else:
    … mem.PyMem_Calloc(…)
```

Reply:
I'm not clear on what the oldest version of CPython that Cython supports is. https://cython.org/ says 3.3 is the oldest, presumably referring to Cython 0.28.5. The closest thing I could find to a statement about CPython version support for new features (which this PR presumably is) is https://github.com/cython/cython/issues/2800#issuecomment-466781420. Do you expect this PR to be merged into Cython 3.0, which maybe will support CPython 3.4, or into Cython 3.1, which maybe will support CPython ≥3.5? If the latter, there's no need for the version guard. If the former, the version guard would need to go around already-existing uses of `PyObject_Calloc`, which don't have a version guard. That seems to me to be beyond the scope of this PR.

Reply:
The guard is only meant for the test. And I would see this as a feature for 3.0, although we made some exceptions like this for 0.29.x already, so I might end up cherry picking it over to the current release branch.

Reply:
Both [`PyMem_Calloc`](https://docs.python.org/3/c-api/memory.html#c.PyMem_Calloc) and [`PyMem_RawCalloc`](https://docs.python.org/3/c-api/memory.html#c.PyMem_RawCalloc) were added in 3.5, so do you actually mean the following?

```diff
diff --git a/tests/run/cpython_capi.pyx b/tests/run/cpython_capi.pyx
index 6d50352b0..f6c57873e 100644
--- a/tests/run/cpython_capi.pyx
+++ b/tests/run/cpython_capi.pyx
@@ -1,6 +1,7 @@
 # mode: run
 # tag: c-api
 
+from cpython.version cimport PY_VERSION_HEX
 from cpython cimport mem
 from cpython.pystate cimport PyGILState_Ensure, PyGILState_Release, PyGILState_STATE
 
@@ -21,13 +22,16 @@ def test_pymalloc():
     3
     """
     cdef short i
-    cdef short* s = <short*> mem.PyMem_Calloc(10, sizeof(short))
-    if not s:
-        raise MemoryError()
-    try:
-        i = _assert_calloc(s, 10)
-    finally:
-        mem.PyMem_Free(s)
+    if PY_VERSION_HEX >= 0x03050000:
+        cdef short* s = <short*> mem.PyMem_Calloc(10, sizeof(short))
+        if not s:
+            raise MemoryError()
+        try:
+            i = _assert_calloc(s, 10)
+        finally:
+            mem.PyMem_Free(s)
+    else:
+        i = 3
     cdef char* m2
     cdef char* m = <char*> mem.PyMem_Malloc(20)
     assert m
@@ -53,13 +57,16 @@ def test_pymalloc_raw():
     cdef char* m
     cdef char* m2 = NULL
     with nogil:
-        s = <short*> mem.PyMem_RawCalloc(10, sizeof(short))
-        if not s:
-            raise MemoryError()
-        try:
-            i = _assert_calloc(s, 10)
-        finally:
-            mem.PyMem_RawFree(s)
+        if PY_VERSION_HEX >= 0x03050000:
+            s = <short*> mem.PyMem_RawCalloc(10, sizeof(short))
+            if not s:
+                raise MemoryError()
+            try:
+                i = _assert_calloc(s, 10)
+            finally:
+                mem.PyMem_RawFree(s)
+        else:
+            i = 3
         m = <char*> mem.PyMem_RawMalloc(20)
         if not m:
             raise MemoryError()
```